### PR TITLE
fix virtual dispatch issues identified by clang-tidy

### PR DESCRIPTION
**Pull request by wjwwood**
fixes https://github.com/space-ros/space-ros/issues/8

~~I might come back and try to de-duplicate some of the code, but I couldn't find a nice solution yet.~~

### CODE REVIEW

**Comment by alsora** [comment]
The changes look good to me.
However, it seems like the link that describes the problem is dead (or I can't access to see it)

**Comment by wjwwood** [comment]
We moved the issue to a new place, which should be accessible. https://github.com/space-ros/space-ros/issues/8

**Comment by wjwwood** [comment]
CI:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/16533/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/11120/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/16907/)

File: rclcpp/include/rclcpp/context.hpp [hunk 1]
```diff
@@ -318,7 +323,6 @@ class Context : public std::enable_shared_from_this<Context>
 
   /// Interrupt any blocking sleep_for calls, causing them to return immediately and return true.
   RCLCPP_PUBLIC
-  virtual
   void
   interrupt_all_sleep_for();
```
Comment (ivanpauno) [inline rclcpp/include/rclcpp/context.hpp:327]: is this related to the PR description?
I don't see directly how.

File: rclcpp/include/rclcpp/context.hpp [hunk 2]
```diff
@@ -351,7 +355,6 @@ class Context : public std::enable_shared_from_this<Context>
   // Called by constructor and destructor to clean up by finalizing the
   // shutdown rcl context and preparing for a new init cycle.
   RCLCPP_PUBLIC
-  virtual
   void
   clean_up();
```
Comment (ivanpauno) [inline rclcpp/include/rclcpp/context.hpp:359]: same here

File: rclcpp/include/rclcpp/node_interfaces/node_base.hpp [hunk 3]
```diff
@@ -39,14 +39,22 @@ class NodeBase : public NodeBaseInterface, public std::enable_shared_from_this<N
...
   RCLCPP_PUBLIC
   NodeBase(
     const std::string & node_name,
     const std::string & namespace_,
     rclcpp::Context::SharedPtr context,
     const rcl_node_options_t & rcl_node_options,
     bool use_intra_process_default,
-    bool enable_topic_statistics_default);
+    bool enable_topic_statistics_default,
+    rclcpp::CallbackGroup::SharedPtr default_callback_group = nullptr);
```
Comment (ivanpauno) [inline rclcpp/include/rclcpp/node_interfaces/node_base.hpp:57]: is this related to the PR?

Comment (ivanpauno) [reply rclcpp/include/rclcpp/node_interfaces/node_base.hpp:57]: It seems okay though, just trying to understand if it's an unrelated addition or it's related to the virtual dispatch in constructor/destructor issue.

File: rclcpp/include/rclcpp/node_interfaces/node_parameters.hpp [hunk 4]
```diff
@@ -84,6 +84,16 @@ class NodeParameters : public NodeParametersInterface
...
+  /// Constructor.
+  /**
+   * If using automatically_declare_parameters_from_overrides, overrides of
+   * get_parameter_overrides(), has_parameter(), declare_parameter() will not
+   * be respected.
+   * If this is an issue, pass false for
+   * automatically_declare_parameters_from_overrides and invoke
+   * perform_automatically_declare_parameters_from_overrides() manually after
+   * construction.
+   */
```
Comment (ivanpauno) [inline rclcpp/include/rclcpp/node_interfaces/node_parameters.hpp:96]: The new `perform_automatically_declare_parameters_from_overrides()` is protected though, does the suggestion apply only when extending `NodeParameters()`?

**Review by ivanpauno — COMMENTED** [review]
There are some changes that doesn't seem directly related to the PR.
I have left some comments in those.
If they aren't related, maybe we can open new PRs for them (?).

Otherwise LGTM!

File: rclcpp/include/rclcpp/context.hpp (hunk 1 above)
Comment (wjwwood) [reply rclcpp/include/rclcpp/context.hpp:327]: These functions were called from the destructor and therefore cannot be virtual.

File: rclcpp/include/rclcpp/node_interfaces/node_base.hpp (hunk 3 above)
Comment (wjwwood) [reply rclcpp/include/rclcpp/node_interfaces/node_base.hpp:57]: It's related, as I need to pass the default callback group in order to avoid the issues with calling virtual methods in the destructor.

File: rclcpp/include/rclcpp/node_interfaces/node_parameters.hpp (hunk 4 above)
Comment (wjwwood) [reply rclcpp/include/rclcpp/node_interfaces/node_parameters.hpp:96]: Yes, only when extending it, which is the case if you were to override one of the aforementioned virtual methods.

**Comment by wjwwood** [comment]
I think I addressed all your comments with just replies, but please have another look when you can.

**Comment by wjwwood** [comment]
CI:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/16675/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/11261/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/17071/)

**Review by fujitatomoya — APPROVED** [review]
lgtm

**Comment by wjwwood** [comment]
New CI after rebase (I think the Windows failure was fixed on master?):

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/16697/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/11280/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/17095/)

**Comment by wjwwood** [comment]
I think I fixed the compiler warnings on Windows (finally), CI:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/16845/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/11423/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/17253/)

**Comment by wjwwood** [comment]
Uncrustify, new CI:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/16853/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/11430/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/17260/)

**Comment by wjwwood** [comment]
The Windows test failure is unrelated and flaky.

File: rclcpp/src/rclcpp/graph_listener.cpp [hunk 5]
```diff
@@ -48,7 +48,7 @@ GraphListener::GraphListener(const std::shared_ptr<Context> & parent_context)
 
 GraphListener::~GraphListener()
 {
-  this->shutdown(std::nothrow);
+  GraphListener::shutdown(std::nothrow);
```
Comment (ivanpauno) [inline rclcpp/src/rclcpp/graph_listener.cpp:51]: should this class have a note in the header file similar to the other ones?

**Review by ivanpauno — APPROVED** [review]

File: rclcpp/src/rclcpp/graph_listener.cpp (hunk 5 above)
Comment (wjwwood) [reply rclcpp/src/rclcpp/graph_listener.cpp:51]: Good point, I updated the docstrings.

**Comment by wjwwood** [comment]
The test timeout in Rpr is unrelated.
